src/Simulation/Native/src/simulator/kernels.hpp

```diff
@@ -144,6 +144,46 @@ bool isclassical(
     return true;
 }
 
+template <class T, class A>
+std::pair<bool, bool> is_classical_or_entangled(
+    std::vector<std::complex<T>, A> const& wfn,
+    std::size_t q,
+    T eps = 100. * std::numeric_limits<T>::epsilon())
+{
+    std::size_t offset = 1ull << q;
+    bool have0 = false;
+    bool have1 = false;
+    bool notentangled = false;
+
+    std::size_t maski = ~(offset - 1);
+#ifndef _MSC_VER
+#pragma omp parallel for schedule(static) reduction(|| : have0, have1, notentangled)
+    for (std::intptr_t i = 0; i < static_cast<std::intptr_t>(wfn.size()); i += 2 * offset)
+        for (std::intptr_t j = 0; j < static_cast<std::intptr_t>(offset); ++j)
+        {
+            bool has0 = std::norm(wfn[i + j]) >= eps;
+            bool has1 = std::norm(wfn[i + j + offset]) >= eps;
+            have0 = have0 || has0;
+            have1 = have1 || has1;
+            notentangled = notentangled || (has0 && has1);
+        }
```

Comment:
FYI (may be a false alarm):
I have a suspicion that during the last-most iteration of the outer loop (for `i`) nothing prevents `i` to be exactly 1 less than `wfn.size()`, see the condition `i < static_cast<std::intptr_t>(wfn.size())`.

In that case the fragments
* `wfn[i + j]`
* `wfn[i + j + offset]`

are likely to read _outside_ of `wfn`.

To provide more guarantees of the code correctness consider changing the condition to something like
> i <= (static_cast<std::intptr_t>(wfn.size()) **- 2 * offset**)

Reply:
I added a comment that hopefully helps clarify the intent of the iteration, but it's worth answering your specific question about bounds. Note that `i` is always incremented by `2 * offset` where `offset` is at minimum 1 when `q` is 0. Combining this with the condition `i < static_cast<std::intptr_t>(wfn.size());` means that `i` will always be at most equal to `wfn.size() - 2` if the wave function has an even number of entries. Because the wave function represents the possible states of the N qubits, which is 2^N, we know the length of the wave function will always be even. In this same `q` equals 0 case, `j` is limited by `j < static_cast<std::intptr_t>(offset)` which means `j` will be at most 0, giving us the two final indices of `i + j` or `wfn.size() - 2 + 0` and `i + j + offset` or `wfn.size() - 2 + 0 + 1`. For larger values of `q` the maximum possible value of `i` is lower and the range for `j` is higher, but the largest index checked is still always be `wfn.size() - 1`, so we know we are safely within the bounds of the array for any target qubit.